solr/core/src/java/org/apache/solr/cluster/maintenance/InactiveShardRemover.java

```diff
@@ -0,0 +1,219 @@
+ * split
+ */
+public class InactiveShardRemover
+    implements ClusterSingleton, ConfigurablePlugin<InactiveShardRemoverConfig> {
+
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  public static final String PLUGIN_NAME = ".inactive-shard-remover";
+
+  static class DeleteActor {
+
+    private final CoreContainer coreContainer;
+
+    DeleteActor(final CoreContainer coreContainer) {
+      this.coreContainer = coreContainer;
+    }
+
+    void delete(final Slice slice) {
+      CollectionAdminRequest.DeleteShard deleteRequest =
+          CollectionAdminRequest.deleteShard(slice.getCollection(), slice.getName());
+      try {
+        SolrResponse response =
+            coreContainer.getZkController().getSolrCloudManager().request(deleteRequest);
+        if (response.getException() != null) {
+          throw response.getException();
+        }
+      } catch (Exception e) {
+        log.warn("An exception occurred when deleting an inactive shard", e);
+      }
+    }
+  }
+
+  private State state = State.STOPPED;
+
+  private final CoreContainer coreContainer;
+
+  private final DeleteActor deleteActor;
+
+  private ScheduledExecutorService executor;
+
+  private long scheduleIntervalSeconds;
+
+  private long ttlSeconds;
+
+  private int maxDeletesPerCycle;
+
+  /** Constructor invoked via Reflection */
+  public InactiveShardRemover(final CoreContainer cc) {
+    this(cc, new DeleteActor(cc));
+  }
+
+  public InactiveShardRemover(final CoreContainer cc, final DeleteActor deleteActor) {
+    this.coreContainer = cc;
+    this.deleteActor = deleteActor;
+  }
+
+  @Override
+  public void configure(final InactiveShardRemoverConfig cfg) {
+    Objects.requireNonNull(cfg, "config must be specified");
+    cfg.validate();
+    this.scheduleIntervalSeconds = cfg.scheduleIntervalSeconds;
+    this.maxDeletesPerCycle = cfg.maxDeletesPerCycle;
+    this.ttlSeconds = cfg.ttlSeconds;
+  }
+
+  @Override
+  public String getName() {
+    return PLUGIN_NAME;
+  }
+
+  @Override
+  public State getState() {
+    return state;
+  }
+
+  @Override
+  public void start() throws Exception {
+    state = State.STARTING;
+    executor = Executors.newScheduledThreadPool(1, new SolrNamedThreadFactory(PLUGIN_NAME));
+    executor.scheduleAtFixedRate(
+        this::deleteInactiveSlices,
+        scheduleIntervalSeconds,
+        scheduleIntervalSeconds,
+        TimeUnit.SECONDS);
+    state = State.RUNNING;
+  }
+
+  @Override
+  public void stop() {
+    if (state == State.RUNNING) {
+      state = State.STOPPING;
+      ExecutorUtil.shutdownNowAndAwaitTermination(executor);
+    }
+    state = State.STOPPED;
+  }
+
+  @VisibleForTesting
+  void deleteInactiveSlices() {
+    final ClusterState clusterState = coreContainer.getZkController().getClusterState();
+    Collection<Slice> inactiveSlices =
+        clusterState.getCollectionsMap().values().stream()
```

Comment:
Such an innocent looking thing but `getCollectionsMap()` has been a CPU performance problem at the 10s of thousands of collections scale.  Fine for now but eventually it'd be nice to have an alternative.  For example, perhaps each node could tend to this inactive shard removing task for its own cores, ignoring all the others.  And/or maybe a shard split schedules this on a pertinent node.  Any way, we have something in-hand here that works, which addresses a gap since the autoscaling framework's removal in Solr 8.